File: Common/Serialization/SerializerCollection.vb
Public Class SerializerCollection
    Inherits OneToOneOrderedDictionary(Of String, Serializer)

    Public Sub New()
        MyBase.New(Function(S) S.Id)
    End Sub

    Friend Sub LockCurrentElements()
        For Each I In Me.Values
            Me.LockedItems.Add(I)
        Next
    End Sub

    Public Overrides Sub Clear()
        Dim State = Me.LockedItems.ToArray()
        MyBase.Clear()
        For Each I In State
            Me.Add(I)
        Next
    End Sub

    Public Overrides Sub RemoveAt(index As Integer)
        Verify.False(Me.LockedItems.Contains(Me.ItemAt(index)), "Serializer is locked.")
        MyBase.RemoveAt(index)
    End Sub

    Public Overrides Function RemoveKey(key As String) As Boolean
        Verify.False(Me.LockedItems.Contains(Me.Item(key)), "Serializer is locked.")
        Return MyBase.RemoveKey(key)
    End Function

    Public Overrides Function [Set](Value As Serializer) As Boolean
        Verify.False(Me.LockedItems.Contains(Value), "Serializer is locked.")
        Return MyBase.Set(Value)
    End Function

    Public Overrides Property ItemAt(index As Integer) As Serializer
        Get
            Return MyBase.ItemAt(index)
        End Get
        Set(value As Serializer)
            Verify.False(Me.LockedItems.Contains(value), "Serializer is locked.")
            MyBase.ItemAt(index) = value
        End Set
    End Property

    Private ReadOnly LockedItems As HashSet(Of Serializer) = New HashSet(Of Serializer)()

End Class

File: Common/Serialization/SerializerCollection.cs
using System.Linq;
using System.Collections.Generic;

namespace Ks
{
    namespace Common
    {
        public class SerializerCollection : OneToOneOrderedDictionary<string, Serializer>
        {
            public SerializerCollection() : base(S => S.Id)
            {
            }

            internal void LockCurrentElements()
            {
                foreach (var I in this.Values)
                    this.LockedItems.Add(I);
            }

            public override void Clear()
            {
                var State = this.LockedItems.ToArray();
                base.Clear();
                foreach (var I in State)
                    this.Add(I);
            }

            public override void RemoveAt(int index)
            {
                Verify.False(this.LockedItems.Contains(this[index]), "Serializer is locked.");
                base.RemoveAt(index);
            }

            public override bool RemoveKey(string key)
            {
                Verify.False(this.LockedItems.Contains(this[key]), "Serializer is locked.");
                return base.RemoveKey(key);
            }

            public override bool Set(Serializer Value)
            {
                Verify.False(this.LockedItems.Contains(Value), "Serializer is locked.");
                return base.Set(Value);
            }

            public override Serializer this[int index]
            {
                get
                {
                    return base[index];
                }
                set
                {
                    Verify.False(this.LockedItems.Contains(value), "Serializer is locked.");
                    base[index] = value;
                }
            }

            private readonly HashSet<Serializer> LockedItems = new HashSet<Serializer>();
        }
    }
}
